lib/run/secure-fs.js

```diff
@@ -0,0 +1,183 @@
+const _ = require('lodash'),
+    path = require('path'),
+    util = require('util'),
+    Readable = require('stream').Readable,
+
+    PPERM_ERR = 'PPERM: insecure file access outside working directory',
+
+    // Use simple character check instead of regex to prevent regex attack
+    /*
+     * Windows root directory can be of the following from
+     *
+     * | File System | Actual           | Modified          |
+     * |-------------|------------------|-------------------|
+     * | LFS (Local) | C:\Program       | /C:/Program       |
+     * | UNC         | \\Server\Program | ///Server/Program |
+     */
+    isWindowsRoot = function (path) {
+        const drive = path.charAt(1);
+
+        return ((path.charAt(0) === '/') &&
+            ((drive >= 'A' && drive <= 'Z') || (drive >= 'a' && drive <= 'z')) &&
+            (path.charAt(2) === ':')) ||
+            path.slice(0, 3) === '///'; // Modified UNC path
+    },
+
+    stripTrailingSep = function (thePath) {
+        if (thePath[thePath.length - 1] === path.sep) {
+            return thePath.slice(0, -1);
+        }
+
+        return thePath;
+    },
+
+    pathIsInside = function (thePath, potentialParent) {
+        // For inside-directory checking, we want to allow trailing slashes, so normalize.
+        thePath = stripTrailingSep(thePath);
+        potentialParent = stripTrailingSep(potentialParent);
+
+        // Node treats only Windows as case-insensitive in its path module; we follow those conventions.
+        if (global.process.platform === 'win32') {
+            thePath = thePath.toLowerCase();
+            potentialParent = potentialParent.toLowerCase();
+        }
+
+        return thePath.lastIndexOf(potentialParent, 0) === 0 &&
+            (
+                thePath[potentialParent.length] === path.sep ||
+                thePath[potentialParent.length] === undefined
+            );
+    };
+
+/**
+ * Secure file resolver wrapper over fs. It only allow access to files inside working directory unless specified.
+ *
+ * @param {*} fs - Node's fs module or module with similar interface
+ * @param {*} workingDir - Path of working directory
+ * @param {*} [insecureFileRead=false] - If true, allow reading files outside working directory
+ * @param {*} [fileWhitelist=[]] - List of allowed files outside of working directory
+ */
+function SecureFs (fs, workingDir, insecureFileRead = false, fileWhitelist = []) {
+    this._fs = fs;
+    this._path = path;
+    this.constants = this._fs.constants;
+
+    this.workingDir = workingDir;
+    this.insecureFileRead = insecureFileRead;
+    this.fileWhitelist = fileWhitelist;
+
+    this.isWindows = global.process.platform === 'win32';
+
+    // Attach all functions in fs to secure-fs except overridden functions
+    Object.getOwnPropertyNames(fs).map((prop) => {
```

Comment:
This is an anti-pattern. This will cause the prototype to be overridden on creating every new instance. The module converts the fs from being stateless to being stateful, hence `fs` here is not an argument. 

cc: @shamasis